pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,628 @@
+			},
+		},
+		resyncPeriod,
+	)
+	return c, nil
+}
+
+func newNodeConsistentHashMap() *consistenthash.Map {
+	return consistenthash.New(defaultVirtualNodeReplicas, defaultHashFn)
+}
+
+func (c *Cluster) filterNodes() (nodes []*corev1.Node, err error) {
+	ipPools, err := c.ipPoolLister.List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "List node selectors of nodes failed")
+		return
+	}
+	nodeSet := sets.String{}
+	for _, ipPool := range ipPools {
+		c.conHashMap[ipPool.Name] = newNodeConsistentHashMap()
+		nodeSel, _ := metav1.LabelSelectorAsSelector(&ipPool.Spec.NodeSelector)
+		nodeList, _ := c.nodeLister.List(nodeSel)
+		for _, node := range nodeList {
+			nodeName := node.Name
+			if !nodeSet.Has(nodeName) {
+				nodeSet.Insert(node.Name)
+				nodes = append(nodes, node)
+			}
+		}
+	}
+	return
+}
+
+func (c *Cluster) syncClusterMembers() (clusterNodes []string) {
+	nodes, err := c.filterNodes()
+	if err != nil {
+		klog.ErrorS(err, "Listing Nodes failed")
+		return
+	}
+	klog.V(3).InfoS("Sync cluster members", "node num", len(nodes))
+
+	for _, node := range nodes {
+		nodeAddr, err := k8s.GetNodeAddr(node)
+		if err != nil {
+			klog.ErrorS(err, "Obtain local IP address from K8s node failed")
+			continue
+		}
+		member := fmt.Sprintf("%s:%d", nodeAddr, c.bindPort)
+		clusterNodes = append(clusterNodes, member)
+	}
+	return
+}
+
+func (c *Cluster) shouldJoinCluster(nodeName string) bool {
+	node, err := c.nodeLister.Get(nodeName)
+	if err != nil {
+		klog.ErrorS(err, "Get node failed", "nodeName", nodeName)
+		return false
+	}
+	return c.matchNodeSelectors(node).Len() > 0
+}
+
+func (c *Cluster) matchNodeSelector(node *corev1.Node, ipPool *v1alpha2.ExternalIPPool) bool {
+	nodeSelector, _ := metav1.LabelSelectorAsSelector(&ipPool.Spec.NodeSelector)
+	if nodeSelector.Matches(labels.Set(node.GetLabels())) {
+		return true
+	}
+	return false
+}
+
+func (c *Cluster) matchNodeSelectors(node *corev1.Node) sets.String {
+	pools := sets.NewString()
+	ipPools, err := c.ipPoolLister.List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "matchNodeSelectors failed")
+		return pools
+	}
+	for _, ipPool := range ipPools {
+		nodeSelector, _ := metav1.LabelSelectorAsSelector(&ipPool.Spec.NodeSelector)
+		if nodeSelector.Matches(labels.Set(node.GetLabels())) {
+			pools.Insert(ipPool.Name)
+		}
+	}
+	return pools
+}
+
+func (c *Cluster) memberNum() int {
+	c.memberRWLock.RLock()
+	defer c.memberRWLock.RUnlock()
+	return c.existingMembers.Len()
+}
+
+func (c *Cluster) addExistingMembers(members ...string) {
+	c.memberRWLock.Lock()
+	defer c.memberRWLock.Unlock()
+	c.existingMembers.Insert(members...)
+	klog.V(2).InfoS("Add existingMember", "memberNum", c.existingMembers.Len())
+}
+
+func (c *Cluster) updateExistingMember(node *corev1.Node, add bool) {
+	c.memberRWLock.Lock()
+	defer c.memberRWLock.Unlock()
+	nodeAddr, err := k8s.GetNodeAddr(node)
+	if err != nil {
+		klog.ErrorS(err, "Failed to obtain local IP address")
```

Comment:
The method is not called for local node only, please rephrase.

Reply:
which method？Cluster.updateExistingMember？

Reply:
Yes, doesn't all node call it?